### PR TITLE
NV18: FEVM: setup FEVM during genesis: create ETH0 actor

### PR DESCRIPTION
**Pull request by vyzo**
we need this for wallaby and devnets, will be created by migration normally.

### CODE REVIEW

**Comment by vyzo** [comment]
CI is decisively not happy; fixing.

**Review by arajasek — APPROVED** [review]
LGTM barring sad CI -- feel free to land after fixing that, or get an FVM-y person to also :eye:

**Comment by vyzo** [comment]
main issue was from,to inversion in the implicit message, fixed that, but we now have an annoying issue in the fvm kernel rejecting our message because origin doesn't have a predictable address... SIGH!!!!!

File: chain/gen/genesis/fevm.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/filecoin-project/go-state-types/network"
+
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/consensus/filcns"
+	"github.com/filecoin-project/lotus/chain/state"
+	"github.com/filecoin-project/lotus/chain/store"
+	"github.com/filecoin-project/lotus/chain/vm"
+)
+
+func SetupFEVM(ctx context.Context, cs *store.ChainStore, sys vm.SyscallBuilder, sroot cid.Cid, nv network.Version) (cid.Cid, error) {
```
Comment (Stebalien) [inline chain/gen/genesis/fevm.go:25]: nit: I'd expect this to also handle setting up the _EAM_, but we can rework that later.

File: chain/gen/genesis/fevm.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	// return
+	// code_begin:
+	// push1 0x00
+	// push1 0x00
+	// return
+	// code_end:
+	initcode, err := hex.DecodeString("600580600b6000396000f360006000f3")
+	if err != nil {
+		return cid.Undef, fmt.Errorf("failed to parse ETH0 init code during genesis: %w", err)
+	}
```
Comment (Stebalien) [inline chain/gen/genesis/fevm.go:83]: for future readers, this will go away very soon.

File: chain/gen/genesis/genesis.go [hunk 3]
```diff
@@ -694,6 +701,7 @@ func SetupEAM(_ context.Context, nst *state.StateTree, nv network.Version) error
 		Code:    codecid,
 		Head:    vm.EmptyObjectCid,
 		Balance: big.Zero(),
+		Address: &builtintypes.EthereumAddressManagerActorAddr, // so that it can create ETH0
```
Comment (Stebalien) [inline chain/gen/genesis/genesis.go:704]: This is very wrong and could break _many_ things.

Comment (Stebalien) [reply chain/gen/genesis/genesis.go:704]: We assume that these addresses are always f1/f3/f4, never f0.

Comment (Stebalien) [reply chain/gen/genesis/genesis.go:704]: I'm going to see what I can do to fix this.

**Comment by Stebalien** [comment]
@vyzo yes, I know. I retconned everything and did it the boring way. Unfortunately, this was the easiest way to avoid setting the predictable address.

**Comment by Stebalien** [comment]
Ok, I undid my changes because genesis has some issues. We're going with this approach, because whatever.

**Review by Stebalien — APPROVED** [review]
Ok, a day of messing with this, LGTM for now.

**Pull request merged**